core/controllers/skill_editor_test.py

```diff
@@ -96,47 +157,134 @@ def test_editable_skill_handler_put(self):
+        with self.swap(feconf, 'ENABLE_NEW_STRUCTURES', False):
+            response = self.testapp.get(self.url, expect_errors=True)
+            self.assertEqual(response.status_int, 404)
+        # Check GET returns 404 when cannot get skill by id.
+        feconf_swap = self.swap(feconf, 'ENABLE_NEW_STRUCTURES', True)
+        skill_services_swap = self.swap(
+            skill_services, 'get_skill_by_id', self._mock_get_skill_by_id)
+        with feconf_swap, skill_services_swap:
+            response = self.testapp.get(self.url, expect_errors=True)
+            self.assertEqual(response.status_int, 404)
+        self.logout()
+
+    def test_editable_skill_handler_put_succeeds(self):
+        self.login(self.ADMIN_EMAIL)
+        with self.swap(feconf, 'ENABLE_NEW_STRUCTURES', True):
+            csrf_token = self._get_csrf_token_for_put()
+            # Check that admins can edit a skill.
             json_response = self.put_json(
-                '%s/%s' % (
-                    feconf.SKILL_EDITOR_DATA_URL_PREFIX, self.skill_id),
-                change_cmd, csrf_token=csrf_token)
+                self.url, self.put_payload, csrf_token=csrf_token)
             self.assertEqual(self.skill_id, json_response['skill']['id'])
             self.assertEqual(
                 'New Description', json_response['skill']['description'])
-            self.logout()
+        self.logout()
 
-            # Check that non-admins cannot edit a skill.
-            self.login(self.NEW_USER_EMAIL)
-            json_response = self.put_json(
-                '%s/%s' % (
-                    feconf.SKILL_EDITOR_DATA_URL_PREFIX, self.skill_id),
-                change_cmd, csrf_token=csrf_token, expect_errors=True,
-                expected_status_int=401)
-            self.assertEqual(json_response['status_code'], 401)
-            self.logout()
+    def test_editable_skill_handler_put_fails(self):
+        self.login(self.ADMIN_EMAIL)
+        with self.swap(feconf, 'ENABLE_NEW_STRUCTURES', True):
+            csrf_token = self._get_csrf_token_for_put()
+            # Check PUT returns 404 when new strutures' pages are not enabled.
+            with self.swap(feconf, 'ENABLE_NEW_STRUCTURES', False):
+                self.put_json(self.url, self.put_payload, csrf_token=csrf_token,
+                              expect_errors=True, expected_status_int=404)
+            # Check PUT returns 404 when cannot get skill by id.
+            get_skill_by_id_swap = self.swap(
+                skill_services, 'get_skill_by_id', self._mock_get_skill_by_id)
+            with get_skill_by_id_swap:
+                self.put_json(self.url, {}, csrf_token=csrf_token,
+                              expect_errors=True, expected_status_int=404)
+            # Check PUT returns 400 when an exception is raised updating the
+            # skill.
+            update_skill_swap = self.swap(
+                skill_services, 'update_skill', self._mock_update_skill)
```

Comment:
Can we name _mock_update_skill in a way that conveys that it raises an exception? Perhaps self._raise_exception might be a good name (since then this line of code conveys more clearly that you're swapping update_skill with a function that raises an exception), but you might have other ideas too.

Reply:
Done.